### PR TITLE
Properly handle requests initiated by ServiceWorkers

### PR DESCRIPTION
**Pull request by kzar**
Requests initiated by ServiceWorkers are given a tabId of -1, but we
still want to block tracking requests that they might make. So let's:
 - Stop ignoring requests with a tabId of -1.
 - Create an opaque Tab Object for those requests, but skip storing it
   for later lookup (since by definition we aren't sure which tab the
   request originated from).
 - Consider the site URL for such requests to be the request
   initiator, when available. Otherwise, the request will always be
   considered to have originated from its own URL and will therefore
   always be considered a first party request.

**Reviewer:** @kdzwinel 

## Steps to test this PR:
1. Check that ServiceWorker initiated requests can be blocked, e.g. by using the test page https://privacy-test-pages.glitch.me/privacy-protections/request-blocking/

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/tab-manager.es6.js [hunk 1]
```diff
@@ -36,6 +40,12 @@ class TabManager {
      * get({tabId: ###});
      */
     get (tabData) {
+        // Opaque 'tab' (e.g. ServiceWorker). Create the Tab Object, but don't
+        // store it for later lookup.
+        if (tabData.tabId === -1) {
```
Comment (kdzwinel) [inline shared/js/background/tab-manager.es6.js:45]: So normal tabs go through "create" and then can be fetched via "get", but SW tabs don't go through "create"?

File: shared/js/background/classes/tab.es6.js [hunk 2]
```diff
@@ -48,15 +48,31 @@ class Tab {
         this.cleanAmpUrl = null
         this.requestId = tabData.requestId
         this.status = tabData.status
-        this.site = new Site(this.url)
+
+        // Consider the site URL to be the request initiator for requests fired
+        // inside opaque 'tabs' (e.g. via ServiceWorkers). Otherwise, consider
+        // the site URL to be the request URL.
+        const requestInitiator = tabData.initiator || tabData.originUrl
```
Comment (kdzwinel) [inline shared/js/background/classes/tab.es6.js:55]: Internally service workers always have a clear origin AFAIK - is our method of determining it reliable? In what cases initiator or originUrl are empty?

sw

**Review by kdzwinel — COMMENTED** [review]
I left some questions in code + I wonder what happens when there are multiple service workers at play - is same tabID messing things up (e.g. overwriting them)?

File: shared/js/background/classes/tab.es6.js (hunk 2 above)
Comment (kzar) [reply shared/js/background/classes/tab.es6.js:55]: I don''t think it is safe to assume the initiator URL is available, because in my experience it isn't always available. For example, perhaps requests initiated by other places will have a tab ID of -1 but no initiator URL. I figure it's better not to throw an exception in those cases, or to convert `undefined` to a URL of `"undefined"` by accident.

File: shared/js/background/tab-manager.es6.js (hunk 1 above)
Comment (kzar) [reply shared/js/background/tab-manager.es6.js:45]: Yea, to be honest I ran the code through a debugger to figure out what happens and what needs to happen. [At the start of handleRequest, we call tabManager.get()](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/e3ebc63/shared/js/background/redirect.es6.js#L89) to fetch the Tab Object for the request. For requests initiated by ServiceWorkers, there won't be a Tab Object and so we need to return a temporary opaque Tab Object if we want tracker blocking to work.

In other words there were three things stopping tracker blocking working for such requests:
1. We explicitly ignored requests with tab ID of -1
2. We had no Tab Object for the "tab" (because its not a tab)
3. When creating an temporary Tab Object, the site URL and tab URL was the same and so the request was always considered first-party

**Comment by kzar** [comment]
> I wonder what happens when there are multiple service workers at play - is same tabID messing things up (e.g. overwriting them)?

Hopefully not, since we create a temporary Tab Object and don't save it for later lookup if the tab ID is -1.

**Review by sammacbeth — APPROVED** [review]
